Rlabkey/R/labkey.webdav.R

```diff
@@ -0,0 +1,230 @@
+    }
+
+    if (!file.exists(localFile)){
+        stop (paste0("File does not exist: ", localFile));
+    }
+
+    url <- labkey.webdav._validateRemoteUrl(baseUrl = baseUrl, folderPath = folderPath, fileSet = fileSet, remoteFilePath = remoteFilePath)
+    options <- labkey.getRequestOptions(method="POST")
+
+    pbody <- upload_file(localFile)
+
+    if (!is.null(.lkdefaults[["debug"]]) && .lkdefaults[["debug"]] == TRUE) {
+        print(paste0("URL: ", url))
+        response <- PUT(url=url, config=options, body=pbody, verbose(data_in=TRUE, info=TRUE, ssl=TRUE))
+    } else {
+        response <- PUT(url=url, config=options, body=pbody)
+    }
+
+    processResponse(response, responseType = "text/plain; charset=utf-8")
+
+    return(TRUE)
+}
+
+labkey.webdav.mkDir <- function(baseUrl=NULL, folderPath, remoteFilePath, fileSet="@files")
+{
+    url <- labkey.webdav._validateRemoteUrl(baseUrl = baseUrl, folderPath = folderPath, fileSet = fileSet, remoteFilePath = remoteFilePath)
+
+    options <- labkey.getRequestOptions(method="POST")
+
+    if (!is.null(.lkdefaults[["debug"]]) && .lkdefaults[["debug"]] == TRUE) {
+        print(paste0("URL: ", url))
+        response <- VERB("MKCOL", url=url, config=options, verbose(data_in=TRUE, info=TRUE, ssl=TRUE))
+    } else {
+        response <- VERB("MKCOL", url=url, config=options)
+    }
+
+    processResponse(response, responseType = "text/plain; charset=utf-8")
+
+    return(TRUE)
+}
+
+labkey.webdav._validateRemoteUrl <- function(baseUrl=NULL, folderPath, remoteFilePath, fileSet="@files")
+{
+    baseUrl=labkey.getBaseUrl(baseUrl);
+
+    ## check required parameters
+    if (missing(baseUrl) || is.null(baseUrl) || missing(folderPath) || missing(fileSet) || missing(remoteFilePath)){
+        stop (paste("A value must be specified for each of baseUrl, folderPath, fileSet, and remoteFilePath"));
+    }
+    
+    ## normalize the folder path
+    folderPath <- encodeFolderPath(folderPath);
+
+    return(paste(baseUrl, "_webdav", folderPath, fileSet, "/", remoteFilePath, sep=""))
+}
+
+labkey.webdav.pathExists <- function(baseUrl=NULL, folderPath, remoteFilePath, fileSet="@files")
+{
+    baseUrl=labkey.getBaseUrl(baseUrl);
+
+    if (missing(baseUrl) || is.null(baseUrl) || missing(folderPath) || missing(remoteFilePath)) {
+        stop (paste("A value must be specified for each of baseUrl, folderPath, fileSet, and remoteFilePath"));
+    }
+    
+    ret <- labkey.webdav.listDir(baseUrl, folderPath, fileSet, remoteFilePath)
+    return(is.null(ret$exception))
+}
+
+labkey.webdav.listDir <- function(baseUrl=NULL, folderPath, remoteFilePath, fileSet="@files")
+{
+    baseUrl=labkey.getBaseUrl(baseUrl);
+
+    url <- labkey.webdav._validateRemoteUrl(baseUrl = baseUrl, folderPath = folderPath, fileSet = fileSet, remoteFilePath = remoteFilePath)
+    url <- paste0(url, "?method=JSON")
+    content <- labkey.post(url, pbody = "", responseType = "text/plain; charset=utf-8")
```

Comment:
Isnt accessing private methods exactly what you highlighted here?  What are you thinking I'd do differently?

Reply:
Sry for not being clear. This is regarding the comment in the original ticket "either a) making the generic labkey.get() and .post() methods public, or b) consider a generic webdav get/put helper function?" 
Apparently it's decided on approach b). I'm just commenting that you can still use labkey.get() and post() without making the changes to Rlabkey package.

Reply:
I dont believe labkey.get() and post() are accessible outside the Rlabkey package though?

Reply:
You can access it using Rlabkey:::labkey.get, not ideal but works...

Reply:
No, I dont think this is right.  Code within Rlabkey can, yes, as this PR is in fact doing in several spots.  You are highlighting code that is using labkey.post.  Code outside Rlabkey cannot, since those are private methods, not exposed outside the package.  If you try to use Rlabkey::labkey.get outside this package you should get an error along the lines of "Error: 'labkey.get' is not an exported object from 'namespace:Rlabkey'".

The original question behind this PR was 'can someone write code that leverages Rlabkey to interact with LabKey servers over webdav'. While in theory someone could just make HTTP requests against a LK server using public R packages, Rlabkey has code to build up the proper headers for interacting with LabKey servers, and there is no point in re-creating that.  Therefore, I saw two ways to approach this: a) expose more basic Rlabkey methods outside the package, like labkey.get(), post(), or getRequestOptions().  The latter is really the most important one.  One potential solution would be to alter the NAMESPACE to export these methods.  b) actually write labkey.webdav methods into Rlabkey.  As you can see, I went with the latter.  With a couple exceptions, these are using the core labkey.post and .get methods.  You did highlight a couple spots where I didnt use labkey.get(), which is b/c capturing the response as a file is different than a simple get, and it still is using getRequestOptions(), which seems like the most important code not to replicate.

Reply:
I see where the confusion is. Rlabkey::labkey.get won't be accessible but Rlabkey:::labkey.get (with 3 of ':'s) should be accessible. 

Even though we could have gone with the option of directly using those private rlabkey methods, I agree the webdav addition to Rlabkey is great and very useful. I'll pull down the changes and do a round of testing.

Reply:
Interesting.  That is strange R would allow that.